proposals/unsigned/unsigned.txt

```diff
@@ -0,0 +1,228 @@
+202x in 2017. It is the sixth item on the DIN national body list for
+inclusion in Fortran 202y.
+
+The use cases can be roughly divided into three classes:
+
+- representing unsigned integer
+- bit operations
+- modular arithmetic (2^n for a datatype with n bits)
+
+The two fundamental designs are:
+
+- adding a dedicated type for each use case with the appropriate
+  behavior of aritmetic operators and intrinsic functions on overflow;
+  the different types can possibly just be different kinds for
+  `unsigned(kind=...)`:
+  * `unsigned`: arithmetic operation overflow do not wraparound, or are
+    possibly not even defined
+  * `bits`: bit operations are defined, arithmetic operations do not
+    wraparound or are not defined
+  * `modular`: arithmetic operations wraparound using modular 2^n
+    arithmetic
+- one `unsigned` type that is used for all three use cases; intrinsic
+  functions are used to implement bit operations, various overflow modes
+  (wraparound, checked, saturated), etc. One must choose some default
+  behavior on arithmetic overflow, discussed below.
+
+There is currently no community nor committee agreement which of the two
+fundamnetal designs to do, nor what the default overflow behavior should
+be for arithmetic operations if we go with the second design.
+
+Consequently, we are proposing to implement the second design with
+undefined behavior for arithmetic overflow, consistent with the existing
+signed integers in Fortran, which allows processors to optionally check
+for overflow. This proposal leaves the door open to later implement
+either the first design, or the second design with defined overflow
+behavior (to wraparound). It is also a subset of features that most
+people seem to agree that we need.
+
+The proposal adds a solution to all three use cases (data
+representation, bit operations, modular arithmetic) that processors can
+implement and users can start using. If later we decide to either add
+dedicated types/kinds `bits` and `modular`, or define default arithmetic
+operators' overflow to wraparound, no existing code will break.
+
+## 1.1. Prior art
+
+At least one Fortran compiler, Sun Fortran, supported unsigned integers.
+Documentation can be found at [Oracle]
+(https://docs.oracle.com/cd/E19205-01/819-5263/aevnb/index.html).
+This proposal borrows heavily from that prior art, without sticking
+to it in all details.
+
+## 1.2 Inputs to this proposal
+
+In addition to the references listed above, the discussion at the
+Fortran proposals site
+https://github.com/j3-fortran/fortran_proposals/issues/2
+influenced this proposal.
+
+
+# 2. Goal
+
+Define a new type, UNSIGNED, with a small set of intrinsic operations
+and intrinsic functions that would satisfy most of the use cases listed
+above.
+
+## 2.1 Value range limitation
+
+An UNSIGNED with n bits has a value range between 0 and 2^n-1.
+(Note that Fortran model integers have values between -2^(n-1)+1 and
+2^(n-1)-1).
```

Comment:
And in practice, the range is $-2^(n-1)$ to $2^(n-1)-1$ with two's-complement interpretation.  Fortran's model permits ones'-complement.